sympy/integrals/manualintegrate.py

```diff
@@ -496,6 +496,12 @@ def parts_rule(integral):
         if isinstance(v, sympy.Integral):
             return
 
+        # Set a limit on the number of times u can be used
+        cachekey = u.xreplace({symbol: _cache_dummy})
+        if _parts_u_cache[cachekey] > 5:
```

Comment:
I tried this with bound 2. It seems to work as well.

Reply:
This only catches infinite recursion from the same u term. I wonder if it can happen even with this logic, from slightly different terms each time. My guess is not currently because it only handles LIATE terms but would that continue to be the case if we extend it (e.g., https://github.com/sympy/sympy/issues/15431).

Reply:
Differentiating functions other than trigonometric / exponential is unlikely to return the integral to its original form. This check is meant for these types only. 

With the more targeted check, I reduced the bound to 2 as suggested by @jksuom, and added a test to make sure that `manualintegrate` is still able to do a bunch of similar integrations by parts.  

I don't know how to prevent manualintegrate from creating an infinite sequence of integrands (which is also possible with other methods, like substitution) without imposing an arbitrary limit on the recursion depth (which would hurt with integrands like `x**n * exp(x)`).